### PR TITLE
FIX: Do not enforce 2FA for users who are using social logins

### PR DESCRIPTION
**Pull request by martin-brennan**
Previously, we did not allow setting `enforce_second_factor` to
`all` if any social/external auth providers were enabled for the
site. This was inconsistent, as you could still set it to `staff`,
and lead to a bad experience where since the users were forced to
enable 2FA, they were no longer able to use their social login.

This commit changes things so that:

* If a user only has social/external auth, they will not have 2FA
  enforced even if it is set to all/staff
* We no longer disallow turning on `enforce_second_factor` if
  social/external auth is configured for the site
* We update the site setting description for `enforce_second_factor`
  to reflect this

This way, site admins can enforce 2FA for people who are using
only local logins, and people using social logins can use
2FA on the external site and not need double-2FA on Discourse.

### CODE REVIEW

**Review by SamSaffron — APPROVED** [review]

**Review by Drenmi — APPROVED** [review]
Do we need to consider any security impact for users who have both social login and local login? If someone has both social login and local login, do we enforce 2FA when using a username and password to login?

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -918,8 +918,9 @@ def redirect_to_login_if_required
     end
   end
 
-  def should_enforce_2fa?
-    disqualified_from_2fa_enforcement = request.format.json? || is_api? || current_user.anonymous?
+  def should_enforce_second_factor?
```
[the remark this answers is not in the document]
Comment (martin-brennan) [reply app/controllers/application_controller.rb:921]: Yes same, I like it when stuff like this matches .e.g the site setting. We switch around between 2fa and second factor quite a lot in the code

File: app/models/user.rb [hunk 2]
```diff
@@ -1515,6 +1515,10 @@ def associated_accounts
     result
   end
 
+  def local_login_only?
+    (associated_accounts.empty? && !single_sign_on_record) || has_any_second_factor_methods_enabled?
```
[the remark this answers is not in the document]
Comment (martin-brennan) [reply app/models/user.rb:1519]: Well, if you turn on 2FA the only way you can login is with local login. So I consider the account local login only at that point.

**Comment by martin-brennan** [comment]
Closing this, it's not actually the way we want to go, we want to allow for 2FA enforcement if a social login is used.

**Pull request closed without merging**